### DISCUSSION TITLE
VolumeOutput

### DISCUSSION THREAD

**Discussion by aqildad-create** [pschatzmann/arduino-audio-tools#904] [Q&A] [answered]
for volume print testing i copy i2s data to VolumeOutput  and below is plotter result . 
from plotter i am confuse why is data continuously oscillating ? even if i loud speak still max amplitude cant exceed more then 33000 . 

now form this result how can we find what the actual volume of  i2s stream was because its oscillating  ? 
pre test : i2sIn   data is correct  ,because audio recorder and many other tests are working . 

I2SStream i2sIn;
VolumeOutput out1;
StreamCopy copier(out1, i2sIn); 

 loop{
  copier.copy() ;
    Serial.println(String(out1.volume()));
}

![image](https://github.com/pschatzmann/arduino-audio-tools/assets/65456998/66866c53-7d70-4a05-a245-9a7a65bb6d46)

**Comment by pschatzmann** [comment]
The VolumeOutput is just determining the maximum sample value that is copied to it. If you don't specify any value in the StreamCopy, the number of samples is 1024.

If the sample rate is too high and is not capturing at least 1 full wave cycle, this number might not be stable. You could try to increase the copy size or decrease the sample rate.

However this looks more that you messed up with the number of channels or bits_per_sample and you did not configure them to match on the input and output side. 

Did you try to print the values with CsvOutput so that you can reproduce the result ?

ps: dont do Serial.println(String(out1.volume())); This is pretty bad because you unnecessarily allocate a String on the heap: just call Serial.println(out1.volume());

**Comment by pschatzmann** [comment, marked as answer]
I committed a correction with a [test case](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-generator-volume/streams-generator-volume.ino). This should resolve your issue...

**Reply by aqildad-create** [reply]
yes  I will down load the latest version and then update you . 

but with previous version  i made a test in which i correct the setup and also print in this way Serial.println(out1.volume());

when i speak amplitude changes but still it is oscillating  

setup()
{
AudioInfo from(32000, 2, 32);
out1.begin(from);
}

below is plotter result
![image](https://github.com/pschatzmann/arduino-audio-tools/assets/65456998/eec5c4de-1c46-4164-afc2-37e070930f13)

**Reply by aqildad-create** [reply]
after latest version i am getting below result . 

![image](https://github.com/pschatzmann/arduino-audio-tools/assets/65456998/70b02366-2a90-47a8-8544-848be2c1111a)

**Comment by pschatzmann** [comment]
Not sure if this looks right or not because I do not know what you were recording: If you record [a constant sine tone](https://onlinetonegenerator.com/) the volume should be quite constant as well...
